consensus/src/transactions.rs

```diff
@@ -40,183 +55,223 @@ mod free;
+        self.prepped_txs.append(&mut txs);
 
-    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        self.database.poll_ready(cx).map_err(Into::into)
+        self
     }
 
-    fn call(&mut self, req: VerifyTxRequest) -> Self::Future {
-        let database = self.database.clone();
+    /// Prepare the transactions and advance to the next step: [`VerificationWanted`].
+    ///
+    /// # [`rayon`]
+    ///
+    /// This function will use [`rayon`] to parallelize the preparation process, so should not be called
+    /// in an async function, unless all the transactions given were already prepared, i.e. [`Self::append_prepped_txs`].
+    pub fn prepare(mut self) -> Result<VerificationWanted, ConsensusError> {
+        if !self.txs.is_empty() {
+            self.prepped_txs.append(
+                &mut self
+                    .txs
+                    .into_par_iter()
+                    .map(new_tx_verification_data)
+                    .collect::<Result<_, _>>()?,
+            );
+        }
 
-        async move {
-            match req {
-                VerifyTxRequest::New {
-                    txs,
-                    current_chain_height,
-                    top_hash,
-                    time_for_time_lock,
-                    hf,
-                } => {
-                    prep_and_verify_transactions(
-                        database,
-                        txs,
-                        current_chain_height,
-                        top_hash,
-                        time_for_time_lock,
-                        hf,
-                    )
-                    .await
-                }
+        Ok(VerificationWanted {
+            prepped_txs: self.prepped_txs,
+        })
+    }
+}
 
-                VerifyTxRequest::Prepped {
-                    txs,
-                    current_chain_height,
-                    top_hash,
-                    time_for_time_lock,
-                    hf,
-                } => {
-                    verify_prepped_transactions(
-                        database,
-                        &txs,
-                        current_chain_height,
-                        top_hash,
-                        time_for_time_lock,
-                        hf,
-                    )
-                    .await
-                }
-            }
+/// The step where the type of verification is decided.
+pub struct VerificationWanted {
+    prepped_txs: Vec<TransactionVerificationData>,
+}
+
+impl VerificationWanted {
+    /// Only semantic verification.
+    ///
+    /// Semantic verification is verification that can done without other blockchain data. The [`HardFork`]
+    /// is technically other blockchain data but excluding it reduces the amount of things that can be checked
+    /// significantly, and it is easy to get compared to other blockchain data needed for contextual validation.
+    pub fn just_semantic(self, hf: HardFork) -> SemanticVerification {
```

Comment:
1/2

nit (I think `only` is a more used term in these contexts)
```suggestion
    /// Only semantic verification.
    ///
    /// Semantic verification is verification that can done without other blockchain data. The [`HardFork`]
    /// is technically other blockchain data but excluding it reduces the amount of things that can be checked
    /// significantly, and it is easy to get compared to other blockchain data needed for contextual validation.
    pub fn only_semantic(self, hf: HardFork) -> SemanticVerification {
```